src/action/auth-mobile.js

```diff
@@ -0,0 +1,202 @@
+/**
+ * @fileOverview action to handle mobile specific authentication
+ * using PINs, TouchID, and KeyStore storage.
+ */
+
+import { PIN_LENGTH } from '../config';
+
+const PIN = 'DevicePin';
+const PASS = 'WalletPassword';
+
+class AuthAction {
+  constructor(store, wallet, nav, SecureStore, Fingerprint, Alert) {
+    this._store = store;
+    this._wallet = wallet;
+    this._nav = nav;
+    this._SecureStore = SecureStore;
+    this._Fingerprint = Fingerprint;
+    this._Alert = Alert;
+  }
+
+  //
+  // PIN actions
+  //
+
+  /**
+   * Initialize the set pin view by resetting input values
+   * and then navigating to the view.
+   * @return {undefined}
+   */
+  initSetPin() {
+    this._store.auth.newPin = '';
+    this._store.auth.pinVerify = '';
+    this._nav.goSetPin();
+  }
+
+  /**
+   * Initialize the pin view by resetting input values
+   * and then navigating to the view.
+   * @return {undefined}
+   */
+  initPin() {
+    this._store.auth.pin = '';
+    this._nav.goPin();
+  }
+
+  /**
+   * Append a digit input to the pin parameter.
+   * @param  {string} options.digit The digit to append to the pin
+   * @param  {string} options.param The pin parameter name
+   * @return {undefined}
+   */
+  pushPinDigit({ digit, param }) {
+    const { auth } = this._store;
+    if (auth[param].length < PIN_LENGTH) {
+      auth[param] += digit;
+    }
+    if (auth[param].length < PIN_LENGTH) {
+      return;
+    }
+    if (param === 'newPin') {
+      this._nav.goSetPinConfirm();
+    } else if (param === 'pinVerify') {
+      this.checkNewPin();
+    } else if (param === 'pin') {
+      this.checkPin();
+    }
+  }
+
+  /**
+   * Remove the last digit from the pin parameter.
+   * @param  {string} options.param The pin parameter name
+   * @return {undefined}
+   */
+  popPinDigit({ param }) {
+    const { auth } = this._store;
+    if (auth[param]) {
+      auth[param] = auth[param].slice(0, -1);
+    } else if (param === 'pinVerify') {
+      this.initSetPin();
+    }
+  }
+
+  /**
+   * Check the PIN that was chosen by the user was entered
+   * correctly twice to make sure that there was no typo.
+   * If everything is ok, store the pin in the keystore and
+   * unlock the wallet.
+   * @return {Promise<undefined>}
+   */
+  async checkNewPin() {
+    const { newPin, pinVerify } = this._store.auth;
+    if (newPin.length !== PIN_LENGTH || newPin !== pinVerify) {
+      this._alert('Incorrect PIN', () => this.initSetPin());
```

Comment:
I liked in the previous commit in this PR where it told the user when the PINs did not match. Maybe we could switch to that error message -- is it possible for them to enter a pin of less than the correct length, since it won't let them go to the next screen otherwise?

Reply:
> I liked in the previous commit in this PR where it told the user when the PINs did not match. Maybe we could switch to that error message

Ok. I changed it back.

> is it possible for them to enter a pin of less than the correct length, since it won't let them go to the next screen otherwise?

Nope. The check only happens when all six bubbles are filled.